BParkingNano/plugins/TrackMerger.cc

```diff
@@ -53,127 +57,120 @@ class TrackMerger : public edm::global::EDProducer<> {
+ // const double dcaSig_probe_;         const double dcaSig_tag_;
+  const double dcaSig_;                  const int trkNormChiMin_;
   const int trkNormChiMax_;
 };
 
+
+
+
+
+
 void TrackMerger::produce(edm::StreamID, edm::Event &evt, edm::EventSetup const &stp) const {
-  //get data  
+
+  //input
   edm::ESHandle<MagneticField> bFieldHandle;
   stp.get<IdealMagneticFieldRecord>().get(bFieldHandle);
-  // edm::Handle<reco::BeamSpot> beamSpotHandle;  
-  // evt.getByToken(beamSpotSrc_, beamSpotHandle);
-  // if ( ! beamSpotHandle.isValid() ) {
-  //   edm::LogError("PFCandProducer") << "No beam spot available from EventSetup" ;
-  // }
-  // reco::BeamSpot beamSpot = *beamSpotHandle;
-
   edm::Handle<pat::PackedCandidateCollection> tracks;
   evt.getByToken(tracksToken_, tracks);
-  
   edm::Handle<pat::PackedCandidateCollection> lostTracks;
   evt.getByToken(lostTracksToken_, lostTracks);
-  
-  edm::Handle<pat::MuonCollection> trgMuon;
-  evt.getByToken(trgMuonToken_, trgMuon);
+  edm::Handle<pat::MuonCollection> trgMuons;
+  evt.getByToken(trgMuonToken_, trgMuons);
+
+  //for lost tracks / pf discrimination
+  unsigned int nTracks = tracks->size();
 
-  int nTracks = tracks->size();
-  int nLostTracks = lostTracks->size();      
-  int totalTracks = nTracks + nLostTracks;
+ // do we need all those ints?
+//  int nLostTracks = lostTracks->size();      
+//  int totalTracks = nTracks + nLostTracks;
+ //same as before comment out untill we are sure 
+//  std::unique_ptr<pat::CompositeCandidateCollection> outTag(new pat::CompositeCandidateCollection());
+
+//ok this was CompositeCandidateCollection 
+  std::unique_ptr<pat::CompositeCandidateCollection> tracks_out(new pat::CompositeCandidateCollection);
+
+
+//correct logic but a bit convoluted -> changing to smthn simpler
+ std::vector<pat::PackedCandidate> totalTracks(*tracks);
+
+ totalTracks.insert(totalTracks.end(),lostTracks->begin(),lostTracks->end());
+
+ 
+ // for loop is better to be range based - especially for large ensembles  
+ for( const pat::PackedCandidate & trk: totalTracks){
+
+   //arranging cuts for speed
+   if(!trk.hasTrackDetails()) continue;
+   if(abs(trk.pdgId()) != 211) continue; //do we want also to keep muons?
+   if(trk.pt() < trkPtCut_ ) continue;
+   if(fabs(trk.eta()) > trkEtaCut_) continue;
+
+   if( (trk.pseudoTrack().normalizedChi2() < trkNormChiMin_ &&
+        trkNormChiMin_>=0 ) ||
+       (trk.pseudoTrack().normalizedChi2() > trkNormChiMax_ &&
+        trkNormChiMax_>0)  )    continue; 
+
+   bool skipTrack=true;
+   GlobalPoint trgvtx;
+   for (const pat::Muon & mu: *trgMuons){
+    //remove tracks inside trg muons jet
+    if(reco::deltaR(trk, mu) < drTrg_Cleaning_ && drTrg_Cleaning_ >0) 
+      continue;
+    //if dz is negative it is deactivated
+    if((fabs(trk.vz() - mu.vz()) > dzTrg_cleaning_ && dzTrg_cleaning_ > 0))
+       continue;
+    skipTrack=false;
+    trgvtx=GlobalPoint(mu.vx(),mu.vy(),mu.vz());
```

Comment:
we should double check which is the vtx associated to the muon.
For example, if is it the PV[0] or a PV chosen based on the muon trajectory.

Reply:
Of course we should but according to the documentation is not necessarily the Highest pT one

Reply:
can you point me to the documentation you are referring to?